Calvin/users crud

claalmve opened the pull request:
_Describe your feature here_

## Related PRs

_Optional - any related PRs you're waiting on, or PRs that will conflict, etc_

## Migrations

_Optional - if you added anything to the database through migration(s)_

## Screenshots

_Required for visual changes_

CC: @kwfk

claalmve committed:
feat(user crud functionality): adding crud functionality for users

claalmve committed:
Merge branch 'main' into calvin/users-CRUD

claalmve committed:
feat(making changes to add next-auth): making changes to add next-auth to NBJC application

claalmve committed:
feat(registration): working on registration and login

claalmve committed:
feat(register/signin): constructing back-end flow for registration and signin

claalmve committed:
Merge branch 'main' into calvin/users-CRUD

claalmve committed:
feat(next-auth integration for user authentication into platform): next auth integration

next auth integration for user authentication into platform, incorporated with new users API +
utilizing tables and functionality provided by next-auth

claalmve commented on pages/register.tsx:
```diff
@@ -0,0 +1,42 @@
+// import { csrfToken, signIn } from 'next-auth/client';
```
This whole page will be replaced by @Sonjajo 's PR, this is just a placeholder for now to test

kwfk replied:
sonja's PR current PR doesn't cover user sign up and sign in

kwfk replied:
this should be at `users/signup`

claalmve committed:
refactor(next-auth): removing extraneous code + files

kwfk commented on interfaces/user.ts:
```diff
@@ -0,0 +1,13 @@
+  emailVerified: Joi.date(),
+  image: Joi.string(),
+  hashed_password: Joi.string(),
+}); // .when('$strict', { is: true, then: Joi.object().and('lat', 'long') });
```
remove comments

kwfk commented on pages/admin/index.tsx:
```diff
@@ -0,0 +1,13 @@
+const AdminDash: React.FC = () => {
```
this file shouldn't be in this PR?

kwfk commented on pages/about.tsx:
```diff
@@ -1,16 +1,18 @@
-  </Layout>
-);
+const AboutPage: React.FunctionComponent = () => {
+  return (
```
remove return so there is no diff

kwfk commented on pages/api/orgs/index.ts:
```diff
@@ -15,7 +15,7 @@ export default async (
 
   const { error, value } = OrganizationSchema.validate(req.body, {
-    context: { strict: true },
+    context: { strict: false },
```
why did this change to false?

kwfk commented on pages/index.tsx:
```diff
@@ -17,6 +17,8 @@ const Map = dynamic(() => import('../components/Map'), {
 });
 
 const Home: React.FC = () => {
+  // const [session, loading] = useSession();
```
remove comment or add context

kwfk commented on tsconfig.json:
```diff
@@ -16,5 +16,9 @@
     "baseUrl": "."
   },
   "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
+  "rules": {
+    "no-use-before-define": "off",
+    "@typescript-eslint/no-use-before-define": ["error"]
+  },
```
instead of changing global rules, use local exceptions

kwfk reviewed (changes requested):
haven't tested yet but here are some changes to make but gj 👍

claalmve committed:
style(next-auth): addressing comments from fred

claalmve committed:
refactor(user signup): moving user signup page + renaming component

claalmve committed:
Merge branch 'main' into calvin/users-CRUD

kwfk commented on pages/api/auth/[...nextauth].ts:
```diff
@@ -0,0 +1,64 @@
+          throw new Error('No account exists');
+        }
+        // Verify that their password matches
+        if (user.hashed_password === hashPassword(credentials.password)) {
```
should be hashedPassword?

claalmve replied:
Oh wait i think that's outdated atm, it should be `hashedPassword` now

claalmve committed:
refactor(prisma + .env): adding .env variable and updating prisma

claalmve committed:
fix(prisma user types): changing attributes from snake_case to camelCase because of prisma update

kwfk commented on pages/api/users/index.ts:
```diff
@@ -1,16 +1,44 @@
+    return res.json(sanitizeUser(newUser));
   } catch (err) {
-    res.status(500).json({ statusCode: 500, message: err.message });
+    return CreateError(500, 'Failed to create user due to repeated email', res); // This is temporarily the only error I can think of to reach this point for
```
what about 'duplicate email' instead

kwfk commented on pages/api/users/index.ts:
```diff
@@ -1,16 +1,44 @@
+    firstName: req.body.firstName,
+    lastName: req.body.lastName,
+    email: req.body.email,
+    hashedPassword: hash(req.body.password), // I'm not sure why this has to be lowercase here
```
what has to be lowercase?

kwfk commented on pages/api/users/index.ts:
```diff
@@ -1,16 +1,44 @@
+  // Constructing new user
+  try {
+    const newUser = await prisma.user.create({
+      data: registerInfo,
```
`registerInfo` should be `value`

kwfk commented on pages/api/users/index.ts:
```diff
@@ -1,16 +1,44 @@
+
+  // Construct payload
+  const registerInfo = {
+    firstName: req.body.firstName,
+    lastName: req.body.lastName,
+    email: req.body.email,
+    hashedPassword: hash(req.body.password), // I'm not sure why this has to be lowercase here
```
make sure to handle where the fields don't exist in body

kwfk replied:
actually, you should schema verify the `req.body` and then change password to hash

kwfk commented on pages/api/users/[id].ts:
```diff
@@ -0,0 +1,106 @@
+): Promise<User | null> => {
+  const { error, value } = UserSchema.validate(body);
+  if (error) {
+    throw error;
```
use `CreateError`

kwfk commented on interfaces/user.ts:
```diff
@@ -0,0 +1,13 @@
+  firstName: Joi.string().when('$strict', { is: true, then: Joi.required() }),
+  lastName: Joi.string().when('$strict', { is: true, then: Joi.required() }),
+  role: Joi.string(),
+  email: Joi.string().when('$strict', { is: true, then: Joi.required() }),
```
there's an email validation in Joi

kwfk commented on interfaces/user.ts:
```diff
@@ -0,0 +1,13 @@
+const schema = Joi.object({
+  firstName: Joi.string().when('$strict', { is: true, then: Joi.required() }),
+  lastName: Joi.string().when('$strict', { is: true, then: Joi.required() }),
+  role: Joi.string(),
```
there's a Joi validation for only a set of items

kwfk commented on interfaces/user.ts:
```diff
@@ -0,0 +1,13 @@
+  email: Joi.string().when('$strict', { is: true, then: Joi.required() }),
+  emailVerified: Joi.date(),
+  image: Joi.string(),
+  hashedPassword: Joi.string(),
```
strong password validation

kwfk commented on interfaces/user.ts:
```diff
@@ -0,0 +1,13 @@
+import Joi from 'joi';
+
+const schema = Joi.object({
+  firstName: Joi.string().when('$strict', { is: true, then: Joi.required() }),
+  lastName: Joi.string().when('$strict', { is: true, then: Joi.required() }),
```
let's capitalize the first name and last name before putting it into the db (might not be through Joi)

kwfk reviewed (changes requested):
some minor changes needed but it seems to work! whoo!!

claalmve committed:
fix(user validation): fixing user entry validations

claalmve committed:
fix(refactor schema.prisma): refactored schema.prisma with main merge

n

claalmve committed:
fix(fixing schema.prisma): fixing schema.prisma by running more migrations

kwfk commented on interfaces/user.ts:
```diff
@@ -0,0 +1,18 @@
+  lastName: Joi.string().when('$strict', { is: true, then: Joi.required() }),
+  role: Joi.string().valid('admin', 'moderator', 'organization'),
+  email: Joi.string()
+    .email() // Not sure what parameters to use here, if at all. Should be ok with default email requirements.
```
put comments for me in Github not in code XD

claalmve replied:
Oh yea I typically leave little comments just in case anyone runs by it in the code in the future hahahaha but i can totally remove

kwfk commented on pages/api/sessions.ts:
```diff
@@ -0,0 +1,16 @@
+  if (session) {
+    res.send({ content: session, message: 'Successfully retrieved session.' });
+  } else {
+    res.send({
```
use the createError method in utils

kwfk commented on pages/api/users/[id].ts:
```diff
@@ -0,0 +1,110 @@
+
+  if (req.method === 'GET') {
+    try {
+      const user = await getUser(userId);
```
add a TODO to remove hashpassword

kwfk replied:
Actually let's return the sanitized user

claalmve committed:
fix(password): password

n

claalmve committed:
fix(user password joi): adding more validation to user password in interfaces with joi

n

claalmve committed:
fix(last name password joi): adding last name error message back

kwfk committed:
Merge branch 'main' into calvin/users-CRUD

kwfk commented on pages/api/users/index.ts:
```diff
@@ -1,16 +1,46 @@
+      500,
+      'Failed to create user due to duplicate email',
+      res
+    ); // This is temporarily the only error I can think of that can reach this point in production.
```
Add a TODO instead

kwfk commented on pages/api/users/[id].ts:
```diff
@@ -0,0 +1,110 @@
+    throw new ValidationError(
+      'Failed to validate user payload for updating.',
+      null,
+      null
```
just throw the error Joi gives

kwfk committed:
fix(auth): minor changes to auth

Make email type an email (only necessary for provided signin page). Change hash_password to
hashPassword. Move SanitizedUser to interfaces/user. Add varsIgnorePattern to eslint no-unused-vars
to not check vars prefixed with underscore.

kwfk reviewed (changes requested):
some minor changes necessary but looking good!
also pushed some minor changes I made myself. take a look at the commit i made.

claalmve committed:
fix(next-auth): addressing fred's comments

n

claalmve committed:
Merge branch 'calvin/users-CRUD' of https://github.com/calblueprint/nbjc into calvin/users-CRUD

claalmve committed:
fix(fixing sessions): fixing sessions from fred comment

kwfk commented on pages/api/sessions.ts:
```diff
@@ -0,0 +1,21 @@
+    });
+  }
+  return CreateError(
+    400,
```
should be 401

kwfk reviewed (approved):
one small change but LGTM

claalmve committed:
refactor(merge in new main): merging in new main

n
